Java string variable representation

Snowball generates invalid Java code for this reduced testcase:

```
externals (stem)
strings (s)
define stem as $s=''
```

This case isn't currently exercised by any of the stemmers we ship - to reproduce, simply replace e.g. `algorithms/arabic.sbl` with the above and run `make java`:

```
$ make java
cd java && javac  org/tartarus/snowball/ext/arabicStemmer.java
org/tartarus/snowball/ext/arabicStemmer.java:26: error: incompatible types: StringBuilder cannot be converted to char[]
            current = S_s;
                      ^
org/tartarus/snowball/ext/arabicStemmer.java:28: error: cannot find symbol
            limit = current.length();
                           ^
  symbol:   method length()
  location: variable current of type char[]
2 errors
make: *** [GNUmakefile:602: java/org/tartarus/snowball/ext/arabicStemmer.class] Error 1
```

The problem looks to have been introduced by the changes to use `char[]` for the current string (caf9dceb26bf8a1c633d4655401b76b144cad92e).  String variables are still `java.lang.StringBuilder`.

I wonder if we should also switch string variables to be `char[]`.  We'd then need to track their lengths for ourselves, so e.g.:

```
    private java.lang.StringBuilder S_s = new java.lang.StringBuilder();
```

might become:

```
    private char[] S_s = new char[8];
    private int L_s = 0;
```

@rmuir You implemented the change to use `char[]` - thoughts on this?

--- a/GNUmakefile
+++ b/GNUmakefile
@@ -121,6 +121,7 @@ RUNTIME_HEADERS  = runtime/api.h \
 		   runtime/header.h
 
 JAVARUNTIME_SOURCES = java/org/tartarus/snowball/Among.java \
+		      java/org/tartarus/snowball/CharArraySequence.java \
 		      java/org/tartarus/snowball/SnowballProgram.java \
 		      java/org/tartarus/snowball/SnowballStemmer.java \
 		      java/org/tartarus/snowball/TestApp.java

--- a/compiler/generator.c
+++ b/compiler/generator.c
@@ -1961,6 +1961,8 @@ extern struct generator * create_generator(struct analyser * a, struct options *
 #ifndef DISABLE_PYTHON
     g->max_label = 0;
 #endif
+    g->java_import_arrays = false;
+    g->java_import_chararraysequence = false;
     return g;
 }
 

--- a/compiler/generator_java.c
+++ b/compiler/generator_java.c
@@ -265,7 +265,7 @@ static void generate_AE(struct generator * g, struct node * p) {
             w(g, p->mode == m_forward ? "limit" : "limit_backward"); break;
         case c_lenof: /* Same as sizeof() for Java. */
         case c_sizeof:
-            writef(g, "~V.length()", p);
+            writef(g, "L~V", p);
             break;
         case c_len: /* Same as size() for Java. */
         case c_size:
@@ -739,12 +739,25 @@ static void generate_rightslice(struct generator * g, struct node * p) {
 
 static void generate_assignto(struct generator * g, struct node * p) {
     write_comment(g, p);
-    writef(g, "~Massign_to(~V);~N", p);
+    writef(g, "~Mif (~V.length < limit) {~N~+", p);
+    writef(g, "~M~V = Arrays.copyOf(current, limit);~N~-", p);
+    writef(g, "~M} else {~N~+", p);
+    writef(g, "~MSystem.arraycopy(current, 0, ~V, 0, limit);~N~-", p);
+    writef(g, "~M}~N", p);
+    writef(g, "~ML~V = limit;~N", p);
+    g->java_import_arrays = true;
 }
 
 static void generate_sliceto(struct generator * g, struct node * p) {
     write_comment(g, p);
-    writef(g, "~Mslice_to(~V);~N", p);
+    writef(g, "~Mslice_check();~N", p);
+    writef(g, "~Mif (~V.length < ket - bra) {~N~+", p);
+    writef(g, "~M~V = Arrays.copyOfRange(current, bra, ket);~N~-", p);
+    writef(g, "~M} else {~N~+", p);
+    writef(g, "~MSystem.arraycopy(current, bra, ~V, 0, ket - bra);~N~-", p);
+    writef(g, "~M}~N", p);
+    writef(g, "~ML~V = ket - bra;~N", p);
+    g->java_import_arrays = true;
 }
 
 static void generate_address(struct generator * g, struct node * p) {
@@ -752,7 +765,12 @@ static void generate_address(struct generator * g, struct node * p) {
     if (b != NULL) {
         write_literal_string(g, b);
     } else {
+        write_string(g, "new CharArraySequence(");
         write_varref(g, p->name);
+        write_string(g, ", L");
+        write_varref(g, p->name);
+        write_string(g, ")");
+        g->java_import_chararraysequence = true;
     }
 }
 
@@ -883,12 +901,15 @@ static void generate_dollar(struct generator * g, struct node * p) {
               "~MSnowballProgram ~B0 = new SnowballProgram(this);~N", p);
 
     ++g->copy_from_count;
-    str_assign(g->failure_str, "copy_from(");
+    str_append_string(g->failure_str, "LS_");
+    str_append_s(g->failure_str, p->name->s);
+    str_append_string(g->failure_str, " = length; copy_from(");
     str_append(g->failure_str, savevar);
     str_append_string(g->failure_str, ");");
     writef(g, "~Mcurrent = ~V;~N"
               "~Mcursor = 0;~N"
-              "~Mlimit = current.length();~N", p);
+              "~Mlength = L~V;~N"
+              "~Mlimit = length;~N", p);
     generate(g, p->left);
     if (!g->unreachable) {
         write_margin(g);
@@ -973,7 +994,8 @@ static void generate_grouping(struct generator * g, struct node * p, int complem
 static void generate_namedstring(struct generator * g, struct node * p) {
     write_comment(g, p);
     g->S[0] = p->mode == m_forward ? "" : "_b";
-    write_failure_if(g, "!(eq_s~S0(~V))", p);
+    write_failure_if(g, "!(eq_s~S0(new CharArraySequence(~V, L~V)))", p);
+    g->java_import_chararraysequence = true;
 }
 
 static void generate_literalstring(struct generator * g, struct node * p) {
@@ -1203,10 +1225,18 @@ static void generate_class_begin(struct generator * g) {
     w(g, g->options->package);
     w(g, ";~N~N");
 
+    if (g->java_import_arrays) {
+        w(g, "import java.util.Arrays;~N~N");
+    }
+
     if (g->analyser->amongs) {
         w(g, "import ");
         w(g, g->options->among_class);
         w(g, ";~N~N");
+    }
+
+    if (g->java_import_chararraysequence) {
+        w(g, "import org.tartarus.snowball.CharArraySequence;~N~N");
     }
 
     if (g->copy_from_count > 0) {
@@ -1323,13 +1353,12 @@ static void generate_members(struct generator * g) {
     for (struct name * q = g->analyser->names; q; q = q->next) {
         switch (q->type) {
             case t_string:
-                w(g, "~Mprivate ");
-                write_string(g, g->options->string_class);
-                write_char(g, ' ');
+                w(g, "~Mprivate char[] ");
                 write_varname(g, q);
-                write_string(g, " = new ");
-                write_string(g, g->options->string_class);
-                w(g, "();~N");
+                w(g, " = new char[8];~N");
+                w(g, "~Mprivate int L");
+                write_varname(g, q);
+                w(g, " = 0;~N");
                 wrote_members = true;
                 break;
             case t_integer:

--- a/compiler/header.h
+++ b/compiler/header.h
@@ -382,6 +382,8 @@ struct generator {
     int keep_count;      /* used to number keep/restore pairs to avoid compiler warnings
                             about shadowed variables */
     int temporary_used;  /* track if temporary variable used (Ada and Pascal) */
+    char java_import_arrays; /* need `import java.util.Arrays;` */
+    char java_import_chararraysequence; /* need `import org.tartarus.snowball.CharArraySequence;` */
 };
 
 /* Special values for failure_label in struct generator. */

--- a/java/org/tartarus/snowball/CharArraySequence.java
+++ b/java/org/tartarus/snowball/CharArraySequence.java
@@ -1,0 +1,36 @@
+package org.tartarus.snowball;
+
+/**
+ * Internal class used by Snowball stemmers
+ */
+public class CharArraySequence implements CharSequence {
+    public CharArraySequence(char[] a, int len) {
+        this.a = a;
+        this.len = len;
+    }
+
+    final public char charAt(int index) {
+        if (index < 0 || index >= len) {
+            throw new StringIndexOutOfBoundsException(index);
+        }
+        return a[index];
+    }
+
+    final public int length() {
+        return len;
+    }
+
+    final public CharSequence subSequence(int start, int end) {
+        // Not needed for how we used CharSequence.
+        throw new UnsupportedOperationException();
+    }
+
+    final public String toString() {
+        if (a == null) return "";
+        return new String(a, 0, len);
+    }
+
+    final private char[] a;
+
+    final private int len;
+}

--- a/java/org/tartarus/snowball/SnowballProgram.java
+++ b/java/org/tartarus/snowball/SnowballProgram.java
@@ -77,7 +77,7 @@ public class SnowballProgram implements Serializable {
     }
 
     // current string
-    private char[] current;
+    protected char[] current;
 
     protected int cursor;
     protected int length;
@@ -424,21 +424,6 @@ public class SnowballProgram implements Serializable {
         int adjustment = replace_s(c_bra, c_ket, s);
         if (c_bra <= bra) bra += adjustment;
         if (c_bra <= ket) ket += adjustment;
-    }
-
-    /* Copy the slice into the supplied StringBuilder */
-    protected void slice_to(StringBuilder s)
-    {
-        slice_check();
-        int len = ket - bra;
-        s.setLength(0);
-        s.append(current, bra, len);
-    }
-
-    protected void assign_to(StringBuilder s)
-    {
-        s.setLength(0);
-        s.append(current, 0, limit);
     }
 
 /*

